new lint: [`should_panic_without_expect`]

y21 opened the pull request:
Closes #10956

changelog: new lint: [`should_panic_without_expect`]

Centri3 commented on clippy_lints/src/attrs.rs:
```diff
@@ -716,6 +756,7 @@ impl EarlyLintPass for EarlyAttributes {
         check_mismatched_target_os(cx, attr);
         check_minimal_cfg_condition(cx, attr);
         check_misused_cfg(cx, attr);
+        check_should_panic_reason(cx, attr);
```
The lint is in the `Attributes` pass but the check is in `EarlyAttributes`.

This should be in `Attributes` instead anyway as [early lints in this file are pre-expansion lints](https://github.com/rust-lang/rust-clippy/blob/ee8a429792eb280d20c5e4f4b1549d1dd3b30335/clippy_lints/src/lib.rs#L370), which have historically had countless problems and are quasi-deprecated in rustc. The implementation should work the same as a late pass lint

y21 replied:
The reason I made it part of the early lint pre-expansion pass is because that makes finding the `#[should_panic]` attribute a lot easier. `#[test]` seems to expand to a `const` item with the `should_panic` attribute part of a field of the struct being initialized in the const item somewhere.
`#[should_panic]` doesn't appear in `Attribute`'s `check_attribute` fn, so I guess we'd have to look for these kinds of const items, which sounds a lot more painful than what it's currently doing. Or am I missing something?

Centri3 replied:
Oh yeah that's a good point actually, also more importantly `#[cfg(test)]` and just `#[test]` will cause it to not appear in the HIR *unless* you're running `cargo test`

So this'll unfortunately need to be a pre-expansion lint. To make sure that `#[cfg(test)}` and all work tests should be added for those

Alexendoo replied:
If it can be done outside of a pre expansion pass it should be, the main pre expansion bug is it not picking up lint levels correctly - https://github.com/rust-lang/rust-clippy/issues/6610, for a lint that you might want to globally allow this is particularly bad

`cargo clippy --tests` (or `--all-targets` etc) can be used to run clippy on tests

y21 replied:
Right, okay, that's pretty bad. This makes it quite a bit more complicated, so I suppose we'll have to look for the expanded `#[test]` const item and extract the `should_panic` field that's somewhere in the `test::TestDescAndFn` struct initializer.
I'm not really sure how I'd then get the span of the `#[should_panic]` attribute though, as I don't think this is anywhere in the HIR after `#[test]` expansion

Centri3 replied:
All I can think of is having a pre expansion pass to get the span of the attribute, then a late pass for the actual linting. Perhaps you can match the functions together through the `testfn` field.

It's unfortunately gonna require a hacky solution no matter what

y21 replied:
> Another thing that's worth a shot is using `get_attrs` on the `DefId` of the function called in `testfn`. _maybe_ that'd have `should_panic`

Oh that's smart, I didn't think of that. I looked at the output of `-Zunpretty=hir-tree` and I can see a `#[should_panic]` attribute with the right span on there. That should work and is probably the least hacky way to make this work

Centri3 commented on clippy_lints/src/attrs.rs:
```diff
@@ -767,6 +808,39 @@ fn check_empty_line_after_outer_attr(cx: &EarlyContext<'_>, item: &rustc_ast::It
+            "#[should_panic] attribute without a reason",
+            "consider specifying the expected panic",
+            r#"#[should_panic(expected="<panic message>")]"#.into(),
+            Applicability::Unspecified
```
nit: This should probably be `HasPlaceholders`

Centri3 commented on clippy_lints/src/attrs.rs:
```diff
@@ -767,6 +808,39 @@ fn check_empty_line_after_outer_attr(cx: &EarlyContext<'_>, item: &rustc_ast::It
+            attr.span,
+            "#[should_panic] attribute without a reason",
+            "consider specifying the expected panic",
+            r#"#[should_panic(expected="<panic message>")]"#.into(),
```
nit: There should probably be a space on either side of the `=`

Alexendoo commented:
I'd say call this `should_panic_without_expect`, I wasn't aware you could supply an expected panic message and the title made me think it would be like a comment

y21 commented:
> I'd say call this `should_panic_without_expect`, I wasn't aware you could supply an expected panic message and the title made me think it would be like a comment

Good point, I renamed the lint and also rewrote it as a late lint now

Alexendoo commented on clippy_lints/src/attrs.rs:
```diff
@@ -518,6 +558,42 @@ impl<'tcx> LateLintPass<'tcx> for Attributes {
+                    && let ExprKind::Call(callee, _) = fn_call_closure.kind
+                    && let ExprKind::Path(QPath::Resolved(_, path)) = callee.kind
+                    && let Some(test_function_did) = path.res.opt_def_id()
+                    && let Some(attribute) = cx.tcx.get_attr(test_function_did, sym::should_panic)
```
If the should_panic attribute remains on the function is there a reason we aren't checking that directly? Rather than finding the function via its test description

y21 replied:
Huh, okay I think I made a mistake when I rewrote this from pre expansion to late lint. It seems like I was wrong: `check_attribute` *does* get called for `#[should_panic]`. It must have been that I forgot to run with `--test`, in which case the function is completely cfg'd out and does not show up at all (so it makes sense that the attribute didn't get visited).
In that case we can make this a whole lot simpler.

Centri3 commented:
Maybe rebase and run `cargo bless`? rustfix is ran automatically now so it may be creating a `fixed` file

Alexendoo commented:
You can add `//@no-rustfix` to prevent that

Alexendoo commented on clippy_utils/src/paths.rs:
```diff
@@ -166,3 +166,5 @@ pub const DEBUG_STRUCT: [&str; 4] = ["core", "fmt", "builders", "DebugStruct"];
 pub const ORD_CMP: [&str; 4] = ["core", "cmp", "Ord", "cmp"];
 #[expect(clippy::invalid_paths)] // not sure why it thinks this, it works so
 pub const BOOL_THEN: [&str; 4] = ["core", "bool", "<impl bool>", "then"];
+#[allow(clippy::invalid_paths, reason = "internal lints do not always know about ::test")]
+pub const TEST_DESC_AND_FN: [&str; 3] = ["test", "types", "TestDescAndFn"];
```
Thanks! r=me with this removed

Alexendoo commented:
👍

@bors r+

bors commented:
:pushpin: Commit b4ab67fd4a7d541ad25c79ac0797c889a9ea6a03 has been approved by `Alexendoo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Alexendoo b4ab67fd4a7d541ad25c79ac0797c889a9ea6a03 -->
<!-- homu: {"type":"Approved","sha":"b4ab67fd4a7d541ad25c79ac0797c889a9ea6a03","approver":"Alexendoo","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit b4ab67fd4a7d541ad25c79ac0797c889a9ea6a03 with merge b7cad50eca78610a0c5f453dc6575df8075778ca...
<!-- homu: {"type":"BuildStarted","head_sha":"b4ab67fd4a7d541ad25c79ac0797c889a9ea6a03","merge_sha":"b7cad50eca78610a0c5f453dc6575df8075778ca"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/5911383414/job/16033875574), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/5911383411/job/16033886457), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/5911383409/job/16033994452)
Approved by: Alexendoo
Pushing b7cad50eca78610a0c5f453dc6575df8075778ca to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Alexendoo","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/5911383409/job/16033994452","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/5911383414/job/16033875574","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/5911383411/job/16033886457"},"merge_sha":"b7cad50eca78610a0c5f453dc6575df8075778ca"} -->
